### DISCUSSION TITLE
FHEM/14_SD_UT.pm | IF - ELSE Schlangen und was passiert eigentlich

### DISCUSSION THREAD

**Discussion by sidey79** [RFD-FHEM/RFFHEM#934] [Ideas]
> > Vielleicht könnt ihr meine Annahme zu dem Fehler bestätigen oder wiederlegen?
> 
> Machen wir doch gern. Den Fehler können wir nicht bestätigen. Zusammenfassend haben wir deine Punkte mal aufgegriffen nachdem wir alles uns angesehen haben.

Ok, seltsam dass das nicht der Fehler war... 

> > ...dass ja auch $protocol berücksichtigt wird und das kann nur einen der Werte haben.
> 
> Das ist falsch.

Missverständnis? Ich meine dass $protocol pro Aufruf nur einen Wert hat und die Variable somit nicht nicht gleich 30 und z.B. 86 sein kann.

> 
> ```
>     ### Westinghouse Buttons_five [P29] ###
>     if (!$def && ($protocol == 29 || $protocol == 30)) {
> 
>     ### Novy_840029 [P86] ###
>     if (!$def && ($protocol == 86 || $protocol == 81)) {
> ```
> 
> Das kommt zu tragen, da manche Protokolle sehr ähnlich sein können und aufgrund der Toleranzen "doppelt anschlagen" bei einem Empfang.

Das wird dann aber separat als eigener Datensatz an die sub übergeben und hat mit einem anderen Durchlauf nichts zu tun oder?

> 
> Verschiedene Remotes das gleiche Prot verwenden.

Ja das verstehe ich grundsätzlich.

> ```
>   if($hlen == 3 || $hlen == 5) {
>     ### Novy_840029 [P86] ###
>     if (!$def && ($protocol == 86 || $protocol == 81)) {
>       $deviceCode = substr($rawData,0,2);
>       $devicedef = 'Novy_840029 ' . $deviceCode;
>       $def = $modules{SD_UT}{defptr}{$devicedef};
>     }
>     ### Novy_840039 [P86] ###
>     if (!$def && ($protocol == 86 || $protocol == 81)) {
>       $deviceCode = substr($rawData,0,2);
>       $devicedef = 'Novy_840039 ' . $deviceCode;
>       $def = $modules{SD_UT}{defptr}{$devicedef};
>     }
>   }
> ```
> 
> > ... 1284-1395 in eigene codeblöcke zu verlagern die mit einem Return beendet werden können.
> 
> Das bringt keine Vorteile weil:
> 
> * $deviceCode = substr($rawData,0,3) unterschiedlich ist
return zum Beenden der IF Abfragen sobald $def einen Definierten wert hat würde an den Abfragen vom Prinzip ja nichts ändern, nur dass keine mehr Getätigt werden, sobald eine Definition gefunden wurde.
Die anderen if Abfragen die mit`if (!def ` anfagen sind dann ja in der Regel immer unwahr.

> * $protocol NICHT immer ein Werte annehmen kann
Jetzt bin ich neugierig, wieso die Variable keinen Wert haben kann?

> * $devicedef unterschiedlich ist je Remote / Typ
> * $deviceCode unterschiedlich ist je Remote / Typ
> 
> > ... ebenfalls in eine kleine sub ausgelagert
> 
> Das bringt keine Vorteile, da dann 2 Varianten der Schreibweise im Modul verwendet werden müssen, da keine Vereinheitlichtung möglich ist.

Hmm, das habe ich nicht näher untersucht was an anderen Stellen so passiert. :(

> Sollte man es dennoch umschreiben, so MEHRT sich der Code mit vielen Subs und es quellt auf. (undurchsichtiger) :-(

Das bringt ja nichts, wenn es undurchsichtig wird. Das Gegenteil war ja meine Absicht.

> > ... Denn ich glaube der Fehler, wieso das mit dem IF und Else nicht geklappt hat, liegt Zeile 1736.
> 
> Nein es lag an der falschen Schreibweise ELSIF.

Ahso, dann wäre zumindest das aufgeklärt.

> Nun hoffe ich, dir ein wenig mehr Licht ins Dunkle gebracht zu haben. Deine Überlegungen sind richtig, aber diese sind hier nicht praktikabel an der Stelle.

Teils Teils. Ich hab aber so ein deja vu effekt. Kann es sein, dass wir die Diskussion schon mal hatten?
So Dunkel meine ich mich an was zu erinnern.
Am Ende habe ich die ganze Komplexität nicht durchblickt und das hat mich dazu gebracht Ideen aufzubringen, wie der Mensch besser erkennen kann was der Code machen soll. Auf Hoher Flugebene ist mir das bekannt, aber in den Details stecke ich nicht. Das wird auch to much, da ich noch an ein paar anderen Dingen schraube. 

Grundsätzlich bin ich aber überzeugt, dass man durch eine andere Schreibweise das gleiche Ergebnis bei lesbarem Code hinbekommen könnte. Letztendlich ist das aber mein Problem und vielleicht nicht eures, wenn ihr mit dem Code so gut klarkommt?

@HomeAutoUser  @elektron-bbs

**Comment by elektron-bbs** [comment]
Die Diskussion über die vielen "if" kommt so ziemlich bei jedem Update des SD-UT-Moduls wieder hoch:-)
Ich habe keine Probleme mit den If-Bedingungen. Für mich ist eher schwieriger zu überblicken, wenn von Sub zu Sub und weiter zur nächsten gesprungen wird.
